### ISSUE TITLE
some code snippets

### ISSUE THREAD

**Issue by asukaminato0721** [open-spaced-repetition/fsrs-benchmark#4]
```rs
#![allow(dead_code)]
use csv::Error;
use serde::Deserialize;

#[derive(Deserialize, Debug)]
struct Record {
    review_time: u64,
    card_id: u8,
    i: u8,
    delta_t: u32,
    review_rating: u32,
    y: u8,
    t_history: Option<String>,
    r_history: Option<String>,
}
fn csv_to_vec_fsrsitem() -> Result<(), Error> {
    let csv = "review_time	card_id	i	delta_t	review_rating	y	t_history	r_history
1613941677126	9	1	0	3	1		
1614174626461	9	2	3	3	1	0	3
1614801170699	9	3	7	2	1	0,3	3,3
1615669584650	9	4	10	2	1	0,3,7	3,3,2
1616285388984	9	5	7	3	1	0,3,7,10	3,3,2,2
1617396262568	9	6	13	4	1	0,3,7,10,7	3,3,2,2,3
1620389209452	9	7	35	4	1	0,3,7,10,7,13	3,3,2,2,3,4
1632595372139	9	8	141	3	1	0,3,7,10,7,13,35	3,3,2,2,3,4,4
1651768065052	9	9	222	4	1	0,3,7,10,7,13,35,141	3,3,2,2,3,4,4,3";

    let mut reader = csv::ReaderBuilder::new()
        .delimiter(b'\t')
        .from_reader(csv.as_bytes());

    for record in reader.deserialize::<Record>() {
        dbg!(&record);
    }

    Ok(())
}

fn main() {
    csv_to_vec_fsrsitem();
}
```

a simple csv to `Vec<FSRSItem>` like function.

**Comment by asukaminato0721** [comment]
rewrite from https://github.com/scikit-learn/scikit-learn/blob/55a65a2fa/sklearn/model_selection/_split.py#L1018

```rs
#[derive(Debug)]
pub struct TimeSeriesSplit {
    n_splits: usize,
    max_train_size: Option<usize>,
    test_size: Option<usize>,
    gap: usize,
}

impl TimeSeriesSplit {
    pub fn new(
        n_splits: usize,
        max_train_size: Option<usize>,
        test_size: Option<usize>,
        gap: usize,
    ) -> Self {
        TimeSeriesSplit {
            n_splits,
            max_train_size,
            test_size,
            gap,
        }
    }

    pub fn split(&self, x: &[[usize; 2]]) -> Vec<(Vec<usize>, Vec<usize>)> {
        let n_samples = x.len();
        let n_folds = self.n_splits + 1;
        let test_size = match self.test_size {
            Some(size) => size,
            None => n_samples / n_folds,
        };

        if n_folds > n_samples {
            panic!("Cannot have number of folds greater than the number of samples.");
        }

        let mut splits = vec![];
        let indices = (0..n_samples).collect::<Vec<_>>();
        let n_splits = self.n_splits;

        for start in (n_samples - n_splits * test_size..n_samples).step_by(test_size) {
            let train_end = start.saturating_sub(self.gap);
            let train_indices: Vec<usize>;
            let test_indices: Vec<usize>;

            if let Some(max_train_size) = self.max_train_size {
                train_indices =
                    indices[train_end.saturating_sub(max_train_size)..train_end].to_vec();
                test_indices = indices[start..start + test_size].to_vec();
            } else {
                train_indices = indices[0..train_end].to_vec();
                test_indices = indices[start..start + test_size].to_vec();
            }

            splits.push((train_indices, test_indices));
        }

        splits
    }
}

fn main() {
    let tscv = TimeSeriesSplit::new(5, None, None, 0);
    let splits = tscv.split(&[[1, 2], [3, 4], [1, 2], [3, 4], [1, 2], [3, 4]]);

    for (i, (train_index, test_index)) in splits.iter().enumerate() {
        println!("Fold {i}:");
        println!("  Train: index={train_index:?}");
        println!("  Test:  index={test_index:?}");
    }
}
```

**Comment by asukaminato0721** [comment]
closed since #5
